eitprocessing/datahandling/intervaldata.py

```diff
@@ -0,0 +1,236 @@
+        to keep the driving pressure that was set before the start of the selectioon. A use case where
+        `partial_inclusion` should be set to `False` is "detected_breaths": you might want to ignore partial breaths
+        that started before or ended after the selected period.
+
+        Note that when selecting by time, the end time is always included in the selection if it exists in the original
+        object.
+
+        Args:
+            start_time: earliest time point to include in the copy
+            end_time: latest time point to include in the copy
+            partial_inclusion: whether to include an interval that contains the start_time or end_time
+            newlabel: new label of the copied object
+        """
+        if partial_inclusion is None:
+            partial_inclusion = self.default_partial_inclusion
+        newlabel = newlabel or self.label
+
+        selection_start = start_time
+        selection_end = end_time
+
+        if selection_start is None and selection_end is None:
+            copy_ = copy.deepcopy(self)
+            if newlabel:
+                copy_.label = newlabel
+            return copy_
+        if selection_start is None:
+            selection_start = self.intervals[0].start_time
+        if selection_end is None:
+            selection_end = self.intervals[-1].end_time
+
+        iter_values = self.values or itertools.repeat(None)
+        interval_value_pairs = zip(self.intervals, iter_values, strict=True)
+
+        filterfunc = partial(
+            self._keep_overlapping,
+            selection_start=selection_start,
+            selection_end=selection_end,
+            keep_partial_overlapping=partial_inclusion,
+        )
+        filtered_pairs = list(filter(filterfunc, interval_value_pairs))
+
+        if len(filtered_pairs):
+            intervals, values = zip(*filtered_pairs, strict=True)
+            mapfun = partial(self._replace_start_end_time, selection_start=selection_start, selection_end=selection_end)
+            intervals = list(map(mapfun, intervals))
+        else:
+            intervals, values = [], []
+
+        return type(self)(
+            label=newlabel,
+            name=self.name,
+            unit=self.unit,
+            category=self.category,
+            derived_from=[*self.derived_from, self],
+            intervals=list(intervals),
+            values=list(values) if self.has_values else None,
+        )
+
+    @staticmethod
+    def _keep_overlapping(
+        item: tuple[Interval, Any],
+        selection_start: float,
+        selection_end: float,
+        keep_partial_overlapping: bool,
+    ) -> bool:
+        """Helper function for filtering overlapping interval-value pairs."""
+        interval, _ = item
+
+        if keep_partial_overlapping:
+            return interval.start_time < selection_end and interval.end_time > selection_start
+
+        return interval.start_time >= selection_start and interval.end_time <= selection_end
```

Comment:
I think this does the same thing without the need for partials, etc. I think this is much easier to parse for humans.
It does contain a loop and a list comprehension rather than the filter you made, so probably less efficient. But I think that improved readability compensates for what is probably not a bottleneck in speed/efficiency.
Note that code below is untested.

Also, can we change nomenclature of all `select_by_time` functions to `selection_start` and `selection_end` rather than doing it on the spot just for this one. I think in general, it creates less confusion to have a given keyword have a single function throughout the codebase. Now `start_time ` can either be part of a `select_by_time` function or of an `Interval`. (Alernatively, we can make it `Interval.interval_start` and `Interval.interval_end`, but that looks a bit silly).

```suggestion
        selection_start = start_time or self.intervals[0].start_time
        selection_end = end_time or self.intervals[-1].end_time
        partial_inclusion = partial_inclusion or self.default_partial_inclusion
        newlabel = newlabel or self.label

        # return copy of self if neither start nor end is selected
        if start_time is None and end_time is None:
            copy_ = copy.deepcopy(self)
            copy_.derived_from.append(self)
            copy_.label = newlabel
            return copy_

        selection_conditions = (selection_start, selection_end, partial_inclusion)
        keep_intervals, keep_indices = [], []
        for i, interval in enumerate(self.intervals):
            if self._keep_overlapping(interval, *selection_conditions):
                keep_intervals.append(self._replace_start_end_time(interval, *selection_conditions))
                keep_indices.append(i)
        keep_values = [self.values[x] for x in keep_indices] if self.has_values else None

        return type(self)(
            label=newlabel,
            name=self.name,
            unit=self.unit,
            category=self.category,
            derived_from=[*self.derived_from, self],
            intervals=keep_intervals,
            values=keep_values,
        )

    @staticmethod
    def _keep_overlapping(
        interval: Interval,
        selection_start: float,
        selection_end: float,
        partial_inclusion: bool,
    ) -> bool:
        """Helper function for filtering overlapping interval-value pairs."""
        if partial_inclusion:
            return interval.start_time <= selection_end and interval.end_time >= selection_start
        return interval.start_time >= selection_start and interval.end_time <= selection_end
```

Reply:
I addressed this in some other comment.